core/internal/leet/reader_test.go

```diff
@@ -0,0 +1,326 @@
+package leet_test
+
+import (
+	"fmt"
+	"io"
+	"path/filepath"
+	"testing"
+	"time"
+
+	tea "github.com/charmbracelet/bubbletea"
+	"github.com/stretchr/testify/require"
+	"github.com/wandb/wandb/core/internal/leet"
+	"github.com/wandb/wandb/core/internal/observability"
+	"github.com/wandb/wandb/core/internal/transactionlog"
+	spb "github.com/wandb/wandb/core/pkg/service_go_proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// Verifies the missing-file path in NewWandbReader. reader.go
+// returns a descriptive error when the file doesn't exist.
+func TestNewWandbReader_MissingFile(t *testing.T) {
+	_, err := leet.NewWandbReader("no/such/file.wandb", observability.NewNoOpLogger())
+	require.Error(t, err)
+}
+
+func TestParseHistory_StepAndMetrics(t *testing.T) {
+	h := &spb.HistoryRecord{Item: []*spb.HistoryItem{
+		{NestedKey: []string{"_step"}, ValueJson: "2"},
+		{NestedKey: []string{"loss"}, ValueJson: "0.5"},
+		{NestedKey: []string{"_runtime"}, ValueJson: "1.2"},
+	}}
+	msg := leet.ParseHistory(h).(leet.HistoryMsg)
+	require.Equal(t, 2, msg.Step)
+	require.Equal(t, 0.5, msg.Metrics["loss"])
+}
+
+func TestReadAllRecordsChunked_HistoryThenExit(t *testing.T) {
+	path := filepath.Join(t.TempDir(), "chunky.wandb")
+
+	// Write a valid header + two records (history, exit).
+	w, err := transactionlog.OpenWriter(path)
+	require.NoError(t, err)
+	h := &spb.HistoryRecord{
+		Item: []*spb.HistoryItem{
+			{NestedKey: []string{"_step"}, ValueJson: "1"},
+			{NestedKey: []string{"loss"}, ValueJson: "0.42"},
+		},
+	}
+	err = w.Write(&spb.Record{RecordType: &spb.Record_History{History: h}})
+	require.NoError(t, err)
+	err = w.Write(&spb.Record{RecordType: &spb.Record_Exit{Exit: &spb.RunExitRecord{ExitCode: 0}}})
+	require.NoError(t, err)
+	_ = w.Close()
+
+	r, err := leet.NewWandbReader(path, observability.NewNoOpLogger())
+	require.NoError(t, err)
+
+	cmd := leet.ReadAllRecordsChunked(r)
+	msg := cmd()
+	batch, ok := msg.(leet.ChunkedBatchMsg)
+	require.True(t, ok)
+	require.False(t, batch.HasMore)
+	require.NotEqual(t, 0, batch.Progress)
+	require.NotEqual(t, 0, batch.Msgs)
```

Comment:
Can the test assert on the exact values? Negated checks aren't very strong, there are many ways something can be not equal to zero. For instance, it could be a `int64(0)`.